server/routerlicious/packages/services/src/mongodb.ts

```diff
@@ -7,138 +7,294 @@ import { assert } from "console";
+        );
     }
 
     public async findOrCreate(query: any, value: T): Promise<{ value: T; existing: boolean; }> {
-        const result = await this.collection.findOneAndUpdate(
-            query,
-            {
-                $setOnInsert: value,
-            },
-            {
-                returnOriginal: true,
-                upsert: true,
-            });
-
-        return result.value
-            ? { value: result.value, existing: true }
-            : { value, existing: false };
+        return runWithRetry(
+            async () => {
+                const result = await this.collection.findOneAndUpdate(
+                    query,
+                    {
+                        $setOnInsert: value,
+                    },
+                    {
+                        returnOriginal: true,
+                        upsert: true,
+                    });
+
+                return result.value
+                    ? { value: result.value, existing: true }
+                    : { value, existing: false };
+            }, // api
+            "MongoCollection.findOrCreate", // caller
+            3, // maxRetries
+            1000, // retryAfterMs
+            {}, // telemetryProperties
+            undefined, // ShouldIgnoreError
+            (e) => this.clientSideRetry && MongoErrorRetryAnalyzer.shouldRetry(e), // ShouldRetry
+        );
     }
 
     private async updateCore(filter: any, set: any, addToSet: any, upsert: boolean): Promise<void> {
-        const update: any = {};
-        if (set) {
-            update.$set = set;
-        }
-
-        if (addToSet) {
-            update.$addToSet = addToSet;
-        }
-
-        const options = { upsert };
-
-        await this.collection.updateOne(filter, update, options);
+        return runWithRetry(
+            async () => {
+                const update: any = {};
+                if (set) {
+                    update.$set = set;
+                }
+
+                if (addToSet) {
+                    update.$addToSet = addToSet;
+                }
+
+                const options = { upsert };
+
+                await this.collection.updateOne(filter, update, options);
+            }, // api
+            "MongoCollection.updateCore", // caller
+            3, // maxRetries
+            1000, // retryAfterMs
+            {}, // telemetryProperties
+            undefined, // ShouldIgnoreError
+            (e) => this.clientSideRetry && MongoErrorRetryAnalyzer.shouldRetry(e), // ShouldRetry
+        );
     }
 
     private async updateManyCore(filter: any, set: any, addToSet: any, upsert: boolean): Promise<void> {
-        const update: any = {};
-        if (set) {
-            update.$set = set;
-        }
-
-        if (addToSet) {
-            update.$addToSet = addToSet;
-        }
-
-        const options = { upsert };
-
-        await this.collection.updateMany(filter, update, options);
+        return runWithRetry(
```

Comment:
Could you also add runWithRetry to mongo calls in this file:
https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/packages/lambdas/src/scribe/checkpointManager.ts

Reply:
Actually I think we should take out all `runWithRetry` that wrap the db layer we have when we have this. The reason is:
If there are some real issue, for example, long time 503, that this root level will retry 3 times that 3 seconds, and caller retry 3 more times, which will lead a 9 seconds delay....

Reply:
Makes sense. So then we need to make that change along with this one.